### PR TITLE
[SYCL][Matrix] Add support for tf32 type using the unified interface

### PR DESCRIPTION
**Pull request by dkhaldi**

### CODE REVIEW

File: sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp
Comment (JackAKirk) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:172]: Are you sure this works? I just imagined that storage type could represent more than one element of size T2. But I do not know the details of the spirv case so just wanted to check this is fine.

File: sycl/test/matrix/matrix-tf32-test.cpp
Comment (JackAKirk) [inline sycl/test/matrix/matrix-tf32-test.cpp:84]: nit, but shouldn't this type of usage of `round_to_tf32` really use `joint_matrix_apply`?

**Comment by JackAKirk** [comment]
I think this should be fine from NVPTX point of view: will wait for confirmation from the cuda llvm-test-suite run.

File: sycl/test/matrix/matrix-tf32-test.cpp
Comment (JackAKirk) [inline sycl/test/matrix/matrix-tf32-test.cpp:2]: Shouldn't this kind of runtime test be placed in llvm-test-suite/SYCL/Matrix instead?

File: sycl/test/matrix/matrix-tf32-test.cpp
Comment (dkhaldi) [reply sycl/test/matrix/matrix-tf32-test.cpp:84]: Yes, I will use joint_matrix_apply for A and leave the loop way for B to show both usages.

File: sycl/test/matrix/matrix-tf32-test.cpp
Comment (dkhaldi) [reply sycl/test/matrix/matrix-tf32-test.cpp:2]: I will update the header to only check compilation

Comment (MrSidims) [reply sycl/test/matrix/matrix-tf32-test.cpp:3]: If we use -emit-llvm and FileCheck shouldn't we add checks for the actual IR generated by the compiler?

File: sycl/test/matrix/matrix-tf32-test.cpp
Comment (dkhaldi) [inline sycl/test/matrix/matrix-tf32-test.cpp:96]: move this to ewo test in llvm-test-suite

Comment (dkhaldi) [reply sycl/test/matrix/matrix-tf32-test.cpp:96]: tests are added here: https://github.com/intel/llvm-test-suite/pull/1677

**Comment by dkhaldi** [comment]
@MrSidims, I am changing the name to __spirv_RoundFToTF32INTEL
Is this what you will be using?

**Review by yubingex007-a11y — APPROVED** [review]
LGTM

**Comment by yubingex007-a11y** [comment]
e2e testcases got approval in https://github.com/intel/llvm-test-suite/pull/1677

**Comment by yubingex007-a11y** [comment]
ping?

**Comment by yubingex007-a11y** [comment]
the testcases fail are not related with this pr i think.

**Comment by yubingex007-a11y** [comment]
ping?

File: sycl/include/sycl/ext/oneapi/matrix/matrix-intel.hpp
Comment (steffenlarsen) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-intel.hpp:73]: Since this is in a very general namespace (`sycl::ext::oneapi::detail`) it would be better to have the name be a little more specific. What is it a helper-trait for?

**Review by steffenlarsen — COMMENTED** [review]
Small naming change request, but otherwise LGTM.

File: sycl/include/sycl/ext/oneapi/matrix/matrix-intel.hpp
Comment (dkhaldi) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-intel.hpp:73]: We can call it something like jm_type_interpretation_helper_trait.

File: sycl/include/sycl/ext/oneapi/matrix/matrix-intel.hpp
Comment (dkhaldi) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-intel.hpp:76]: remove fill_argument_type as it is currently unused

File: sycl/include/sycl/ext/oneapi/matrix/matrix-intel.hpp
Comment (steffenlarsen) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-intel.hpp:73]: Sounds good to me! Definitely seems to remove most ambiguity and decreases the chance of name conflicts.

**Review by steffenlarsen — APPROVED** [review]
LGTM!

**Review by JackAKirk — APPROVED** [review]

**Comment by yubingex007-a11y** [comment]
@steffenlarsen Hi, can you help us merge this patch

**Comment by steffenlarsen** [comment]
@yubingex007-a11y - Could you please push another merge commit to rerun pre-commit tests? CI issues should hopefully have been addressed.

File: sycl/test-e2e/Matrix/element_wise_all_ops_tf32.cpp
Comment (Nuullll) [inline sycl/test-e2e/Matrix/element_wise_all_ops_tf32.cpp:10]: Should be `-DSYCL_EXT_ONEAPI_MATRIX_VERSION=4`

File: sycl/test-e2e/Matrix/joint_matrix_tf32.cpp
Comment (Nuullll) [inline sycl/test-e2e/Matrix/joint_matrix_tf32.cpp:10]: Should be `-DSYCL_EXT_ONEAPI_MATRIX_VERSION=4`

File: sycl/test-e2e/Matrix/joint_matrix_tf32_impl.hpp
Comment (Nuullll) [inline sycl/test-e2e/Matrix/joint_matrix_tf32_impl.hpp:77]: No known conversion from `wi_element<>` to `float &`

**Pull request merged**
